libraries/automaton/automaton.lua

```diff
@@ -0,0 +1,204 @@
+        command.input('/ja activate <me>', 'user')
+    end
+}
+
+ftype.fields.remove_all = {
+    data = function(_)
+        local pay_load = {
+            job = 0x12,
+            is_sub = player.sub_job == 0x12,
+
+            item_index = 0,
+            slots = {}
+        }
+
+        for k, v in pairs(data.attachments) do
+            pay_load.slots[k + 1] = v.id - attachment_offset
+        end
+
+        packet.outgoing[0x102][0x12]:inject(pay_load)
+    end,
+}
+
+ftype.fields.equip_head = {
+    data = function(data, head)
+        if head ~= data.head.name then
+            local head_id = get_assembly_id(head)
+            if player.sub_job_id ~= 0x12 and player.main_job_id ~= 0x12 then
+                error('Player\'s job is not Puppetmaster!')
+            end
+
+            if head_id == nil then
+                error(head .. ' is an invalid head!')
+            end
+
+            if not data.available_heads[head_id] then
+                error(head .. ' is not availible!')
+            end
+
+            packet.outgoing[0x102][0x12]:inject({
+                job = 0x12,
+                is_sub = player.sub_job == 0x12,
+
+                item_index = head_id,
+                head = head_id
+            })
+        end
+    end,
+}
+
+ftype.fields.equip_frame = {
+    data = function(data, frame)
+        if frame ~= data.frame.name then
+            local frame_id = get_assembly_id(frame)
+            if player.sub_job_id ~= 0x12 and player.main_job_id ~= 0x12 then
+                error('Player\'s job is not Puppetmaster!')
+            end
+
+            if frame_id == nil then
+                error(frame .. ' is an invalid frame!')
+            end
+
+            local available_frame_offset = 32
+            if not data.available_frames[frame_id - available_frame_offset] then
+                error(frame .. ' is not availible!')
+            end
+
+            packet.outgoing[0x102][0x12]:inject({
+                job = 0x12,
+                is_sub = player.sub_job == 0x12,
+
+                item_index = frame_id,
+                frame = frame_id
+            })
+        end
+    end,
+}
+
+ftype.fields.equip_attachment = {
+    data = function(data, slot, attachment)
+        local attachment_id = get_attachment_id(attachment)
+        if player.sub_job_id ~= 0x12 and player.main_job_id ~= 0x12 then
+            error('Player\'s job is not Puppetmaster!')
+        end
+
+        if attachment_id == nil then
+            error(attachment .. ' is an invalid attachment!')
+        end
+
+        if not data.available_attachments[attachment_id] then
+            error(attachment .. ' is not availible!')
+        end
+
+        local payload = {
+            job = 0x12,
+            is_sub = player.sub_job == 0x12,
+
+            item_index = attachment_id,
+            slots = {}
+        }
+
+        payload.slots[slot] = attachment_id;
+        packet.outgoing[0x102][0x12]:inject(payload)
+    end,
+}
```

Comment:
All three equipping functions are extremely similar and I would extract a lot of code. Could have a function like this:
```lua
local prepare_payload = function(gear, retrieve, available, name)
    -- validate job

    local id = retrieve(gear:normalize())
    -- validate id

    if not available[id] then --[[ error() ]] end

    return {
        -- common payload data
    }
end

-- later
local payload = prepare_payload(attachment, get_attachment_id, data.available_attachments, 'attachment')
payload.slots[slot] = payload.item_index
packet.outgoing[0x102][0x12]:inject(payload)
```

Something along those lines, sure I missed something, but you get the idea.

Reply:
I pushed a change for this section, I believe I implemented the refactoring you were suggesting.